### ISSUE TITLE
JS-Controller 5.0.12: Adapter is crashing because of a wrong pattern

### ISSUE THREAD

**Issue by RP70DP** [iobroker-community-adapters/ioBroker.mqtt-client#233]
JS-Controller: 5.0.12
Adapter: 1.6.4

```
2023-09-14 23:26:08.910  - [32minfo[39m: host.iobroker instance system.adapter.mqtt-client.0 started with pid 3149
2023-09-14 23:26:11.236  - [32minfo[39m: mqtt-client.0 (3149) Plugin sentry Sentry Plugin disabled for this process because sending of statistic data is disabled for the system
2023-09-14 23:26:11.348  - [32minfo[39m: mqtt-client.0 (3149) starting. Version 1.6.4 in /opt/iobroker/node_modules/iobroker.mqtt-client, node: v18.16.0, js-controller: 5.0.12
2023-09-14 23:26:12.408  - [32minfo[39m: mqtt-client.0 (3149) Try to connect to mqtt://admin:*******************@192.168.0.6:1883?clientId=mqtt-client.0.iobroker
2023-09-14 23:26:12.625  - [31merror[39m: mqtt-client.0 (3149) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2023-09-14 23:26:12.631  - [32minfo[39m: mqtt.0 (456) Client [mqtt-client.0.iobroker] connected with secret 1694726772485_8486
2023-09-14 23:26:12.626  - [31merror[39m: mqtt-client.0 (3149) unhandled promise rejection: The pattern "io.mqtt-client.0.stat.� 0� stat.tasmota_95F" is not a valid ID pattern
2023-09-14 23:26:12.628  - [31merror[39m: mqtt-client.0 (3149) Error: The pattern "io.mqtt-client.0.stat.� 0� stat.tasmota_95F" is not a valid ID pattern
    at Object.pattern2RegEx (/opt/iobroker/node_modules/@iobroker/js-controller-common/build/lib/common/tools.js:2186:15)
    at StateRedisClient.subscribe (/opt/iobroker/node_modules/@iobroker/db-states-redis/build/lib/states/statesInRedisClient.js:865:100)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2023-09-14 23:26:12.628  - [31merror[39m: mqtt-client.0 (3149) The pattern "io.mqtt-client.0.stat.� 0� stat.tasmota_95F" is not a valid ID pattern
2023-09-14 23:26:12.654  - [32minfo[39m: mqtt-client.0 (3149) connected to broker
2023-09-14 23:26:12.745  - [32minfo[39m: mqtt.0 (456) Client [mqtt-client.0.iobroker] subscribes on "mqtt.0.tele.tasmota_85D39A.STATE"
2023-09-14 23:26:13.199  - [32minfo[39m: mqtt-client.0 (3149) terminating
2023-09-14 23:26:13.202  - [33mwarn[39m: mqtt-client.0 (3149) Terminated (UNCAUGHT_EXCEPTION): Without reason
.
.
.
2023-09-14 23:26:14.110  - [32minfo[39m: mqtt.0 (456) Client [mqtt-client.0.iobroker] publishOnSubscribe send all known states
2023-09-14 23:26:14.158  - [32minfo[39m: mqtt.0 (456) Client [mqtt-client.0.iobroker] connection closed: disconnected
2023-09-14 23:26:14.366  - [32minfo[39m: mqtt-client.0 (3149) terminating with timeout
2023-09-14 23:26:14.934  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_341527.LWT: DB closed
2023-09-14 23:26:14.935  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_A36501.STATUS6: DB closed
2023-09-14 23:26:14.936  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_9E62C6.STATUS: DB closed
2023-09-14 23:26:14.936  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_56D73B.INFO3: DB closed
2023-09-14 23:26:14.937  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_A36501.LWT: DB closed
2023-09-14 23:26:14.938  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_3F8AD6.LWT: DB closed
2023-09-14 23:26:14.938  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_85CD09.STATUS: DB closed
2023-09-14 23:26:14.939  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_A36501.POWER2: DB closed
2023-09-14 23:26:14.940  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_95FE9E.STATE: DB closed
2023-09-14 23:26:14.941  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_7E7404.STATUS4: DB closed
2023-09-14 23:26:14.942  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_990526.UPGRADE: DB closed
2023-09-14 23:26:14.942  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_4CC022.STATUS4: DB closed
2023-09-14 23:26:14.943  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_63E60D.RESULT: DB closed
2023-09-14 23:26:14.944  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_4E489B.STATUS11: DB closed
2023-09-14 23:26:14.944  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_53EAB0.STATUS6: DB closed
2023-09-14 23:26:14.945  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_4E4778.STATUS4: DB closed
2023-09-14 23:26:14.945  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_9E62C6.INFO2: DB closed
2023-09-14 23:26:14.946  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tasmota.discovery.840D8E5E4526.config: DB closed
2023-09-14 23:26:14.947  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_BB868C.STATUS3: DB closed
2023-09-14 23:26:14.948  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_990210.INFO3: DB closed
2023-09-14 23:26:14.948  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_BEB72E.STATUS4: DB closed
2023-09-14 23:26:14.949  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_A36501.STATE: DB closed
2023-09-14 23:26:14.949  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_FD2178.STATE: DB closed
2023-09-14 23:26:14.950  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_5E4526.STATUS4: DB closed
2023-09-14 23:26:14.951  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.tele.tasmota_3F8AD6.SENSOR: DB closed
2023-09-14 23:26:14.952  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_95FE9E.STATUS4: DB closed
2023-09-14 23:26:14.953  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_53EAB0.STATUS1: DB closed
2023-09-14 23:26:14.953  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_41CB86.POWER: DB closed
2023-09-14 23:26:14.954  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_59618E.STATUS7: DB closed
2023-09-14 23:26:14.954  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_3F8C42.RESULT: DB closed
2023-09-14 23:26:14.955  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_3F8C42.STATUS6: DB closed
2023-09-14 23:26:14.955  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_990576.STATUS4: DB closed
2023-09-14 23:26:14.956  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_3F8AD6.UPGRADE: DB closed
2023-09-14 23:26:14.956  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_95F446.STATUS4: DB closed
2023-09-14 23:26:14.957  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_41CB86.STATUS2: DB closed
2023-09-14 23:26:14.958  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_53EAB0.UPGRADE: DB closed
2023-09-14 23:26:14.963  - [33mwarn[39m: mqtt-client.0 (3149) DB closed
2023-09-14 23:26:14.964  - [33mwarn[39m: mqtt-client.0 (3149) Could not perform strict object check of state mqtt-client.0.stat.tasmota_98FD5F.STATUS10: DB closed
2023-09-14 23:26:15.262  - [31merror[39m: mqtt-client.0 (3149) Cannot check object existence of "mqtt-client.0.info.connection": Connection is closed.
2023-09-14 23:26:15.271  - [32minfo[39m: mqtt-client.0 (3149) terminating
2023-09-14 23:26:15.393  - [31merror[39m: host.iobroker Caught by controller[0]: [ioredis] Unhandled error event: Error: write EPIPE
2023-09-14 23:26:15.394  - [31merror[39m: host.iobroker Caught by controller[0]:     at afterWriteDispatched (node:internal/stream_base_commons:160:15)
2023-09-14 23:26:15.394  - [31merror[39m: host.iobroker Caught by controller[0]:     at writeGeneric (node:internal/stream_base_commons:151:3)
2023-09-14 23:26:15.394  - [31merror[39m: host.iobroker Caught by controller[0]:     at Socket._writeGeneric (node:net:930:11)
2023-09-14 23:26:15.394  - [31merror[39m: host.iobroker Caught by controller[0]:     at Socket._write (node:net:942:8)
2023-09-14 23:26:15.395  - [31merror[39m: host.iobroker Caught by controller[0]:     at writeOrBuffer (node:internal/streams/writable:392:12)
2023-09-14 23:26:15.395  - [31merror[39m: host.iobroker Caught by controller[0]:     at _write (node:internal/streams/writable:333:10)
2023-09-14 23:26:15.395  - [31merror[39m: host.iobroker Caught by controller[0]:     at Writable.write (node:internal/streams/writable:337:10)
2023-09-14 23:26:15.395  - [31merror[39m: host.iobroker Caught by controller[0]:     at Redis.sendCommand (/opt/iobroker/node_modules/ioredis/built/redis/index.js:679:33)
2023-09-14 23:26:15.396  - [31merror[39m: host.iobroker Caught by controller[0]:     at Redis.publish (/opt/iobroker/node_modules/ioredis/built/commander.js:122:25)
2023-09-14 23:26:15.396  - [31merror[39m: host.iobroker Caught by controller[0]:     at StateRedisClient.setState (/opt/iobroker/node_modules/@iobroker/db-states-redis/build/lib/states/statesInRedisClient.js:610:35)
2023-09-14 23:26:15.396  - [31merror[39m: host.iobroker instance system.adapter.mqtt-client.0 terminated with code 6 (UNCAUGHT_EXCEPTION)
2023-09-14 23:26:15.396  - [32minfo[39m: host.iobroker Restart adapter system.adapter.mqtt-client.0 because enabled
```

**Comment by Apollon77** [comment]
This more seems that there are devices created with some very strange characters

**Comment by RP70DP** [comment]
Right, but the that should not crash the adapter. MQTT Client/Server Adapter don't crash:

![mqtt](https://github.com/iobroker-community-adapters/ioBroker.mqtt-client/assets/62850183/0e69ede2-2961-4ff1-92dc-02db61883489)

**Comment by Apollon77** [comment]
Sure, it iseems that the adapter has an handling to make sure such broken IDs are not created ... but it seems that this logic is not applied when it subscribes for changes because it tries to subscribe to "io.mqtt-client.0.stat.� 0�� �stat.tasmota_95F" ... so yes someone need tolook ito it and fix it

**Comment by foxriver76** [comment]
Please check if it is no longer crashing in 1.6.5

**Comment by RP70DP** [comment]
Thanks a lot for your work!

Adapter isn't crashing anymore but changing the path as mentioned in the log don't work:

`2023-09-28 14:13:51.379  - [32minfo[39m: host.iobroker "system.adapter.mqtt-client.0" enabled
2023-09-28 14:13:51.519  - [32minfo[39m: host.iobroker instance system.adapter.mqtt-client.0 started with pid 26661
2023-09-28 14:13:53.834  - [32minfo[39m: mqtt-client.0 (26661) Plugin sentry Sentry Plugin disabled for this process because sending of statistic data is disabled for the system
2023-09-28 14:13:53.950  - [32minfo[39m: mqtt-client.0 (26661) starting. Version 1.6.5 (non-npm: iobroker-community-adapters/ioBroker.mqtt-client) in /opt/iobroker/node_modules/iobroker.mqtt-client, node: v18.17.1, js-controller: 5.0.12
2023-09-28 14:13:55.001  - [32minfo[39m: mqtt-client.0 (26661) Try to connect to mqtt://admin:*******************@192.168.0.6:1883?clientId=mqtt-client.0.iobroker
2023-09-28 14:13:55.211  - [32minfo[39m: mqtt.0 (559) Client [mqtt-client.0.iobroker] connected with secret ***
2023-09-28 14:13:55.235  - [32minfo[39m: mqtt-client.0 (26661) connected to broker
.
.
.
2023-09-28 14:13:56.752  - [32minfo[39m: mqtt.0 (559) Client [mqtt-client.0.iobroker] publishOnSubscribe send all known states
2023-09-28 14:13:59.129  - [32minfo[39m: mqtt.0 (559) Client [mqtt-client.0.iobroker] subscribes on "mqtt.0.stat._0_stat.tasmota_95F"
2023-09-28 14:13:59.235  - [31merror[39m: mqtt-client.0 (26661) Cannot subscribe to "mqtt-client.0.stat.� 0� stat.tasmota_95F": The pattern "io.mqtt-client.0.stat.� 0� stat.tasmota_95F" is not a valid ID pattern
2023-09-28 14:13:59.333  - [32minfo[39m: mqtt.0 (559) Client [mqtt-client.0.iobroker] publishOnSubscribe
2023-09-28 14:13:59.841  - [33mwarn[39m: mqtt-client.0 (26661) Used invalid characters: mqtt-client.0.stat.� 0� stat.tasmota_95F changed to mqtt-client.0.stat._0_stat.tasmota_95F
2023-09-28 14:13:59.844  - [33mwarn[39m: mqtt-client.0 (26661) Used invalid characters: mqtt-client.0.stat.� 0� stat.tasmota_95F changed to mqtt-client.0.stat._0_stat.tasmota_95F
2023-09-28 14:13:59.907  - [33mwarn[39m: mqtt-client.0 (26661) State "mqtt-client.0.stat._0_stat.tasmota_95F" has no existing object, this might lead to an error in future versions
2023-09-28 14:13:59.909  - [33mwarn[39m: mqtt-client.0 (26661) State "mqtt-client.0.stat._0_stat.tasmota_95F" has no existing object, this might lead to an error in future versions
2023-09-28 14:14:00.769  - [31merror[39m: admin.0 (267) Invalid pattern on subscribe: The pattern "mqtt-client.0.stat.� 0� stat.tasmota_95F" is not a valid ID pattern
2023-09-28 14:14:18.838  - [33mwarn[39m: mqtt.0 (559) Client [mqtt-client.0.iobroker] Message 359 deleted after 11 retries`


![jhghvjhvj](https://github.com/iobroker-community-adapters/ioBroker.mqtt-client/assets/62850183/770b66e0-392c-414c-bdae-886dd8a96b94)

(I deleted the path before restart of the instance)

**Comment by foxriver76** [comment]
if you delete the channel shown in the screenshot will it be created again? Until now I just prevented the crash case, no further adaption yet. But as far as I know it should not be possible to create such objects for a very long time now

**Comment by RP70DP** [comment]
"if you delete the channel shown in the screenshot will it be created again?"

Exactly!

**Comment by RP70DP** [comment]
Done with js-controller 5.0.13.

**Comment by mcm1957** [comment]
Lets keep opne for now.
Looks like adapter does not explicitly filter / replace invalid characters in ids

**Comment by mcm1957** [comment]
fixed with 1.6.5
